server/internal/entity_internal/lightning.go

```diff
@@ -0,0 +1,84 @@
+package entity_internal
+
+import (
+	"github.com/df-mc/dragonfly/server/block"
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/entity"
+	"github.com/df-mc/dragonfly/server/entity/physics"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/go-gl/mathgl/mgl64"
+	"math"
+	"math/rand"
+)
+
+//go:linkname world_performThunder github.com/df-mc/dragonfly/server/world.performThunder
+//noinspection ALL
+var world_performThunder func(w *world.World, pos world.ChunkPos)
+
+func init() {
+	world_performThunder = func(w *world.World, pos world.ChunkPos) {
+		LCG := int32(w.GetUpdateLGC() >> 2)
+
+		chunkX, chunkZ := pos.X(), pos.Z()
+		vec := adjustPosToNearbyEntities(w, mgl64.Vec3{float64(chunkX + (LCG & 0xf)), 0, float64(chunkZ + (LCG >> 8 & 0xf))})
+
+		blockType := w.Block(cube.Pos{int(math.Floor(vec.X())) & 0xf, int(math.Floor(vec.Y())), int(math.Floor(vec.Z())) & 0xf})
+
+		_, tallGrass := blockType.(block.TallGrass)
+		_, flowingWater := blockType.(block.Water)
+		if !tallGrass && !flowingWater {
+			vec = vec.Add(mgl64.Vec3{0, 1, 0})
+		}
+
+		w.AddEntity(entity.NewLightning(vec))
+	}
+}
+
+func canBlockSeeSky(w *world.World, pos mgl64.Vec3) bool {
```

Comment:
This isn't an accurate name as blocks that can see the sky through glass don't matter. This should probably be called `highestBlock`.